test/mpi/dtpools/src/dtpools.c

```diff
@@ -0,0 +1,500 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *
+ *  (C) 2018 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#include <limits.h>
+#include "dtpools_internal.h"
+
+#define DTPI_MAX_FACTOR (64)
+
+/* get list of MPI datatypes at compilation time */
+MPI_Datatype DTP_Basic_type[] = { DTP_MPI_DATATYPE };
+
+/* pointer to create functions for datatypes */
+static DTPI_Creator creators[DTPI_OBJ_LAYOUT_LARGE__NUM];
+
+/* return factor of count */
+static int DTPI_Get_max_fact(int count)
+{
+    int i = 2;
+
+    if ((count / 2) < DTPI_MAX_FACTOR && (count / 2) > 0) {
+        i = (count / 2) + 1;
+    } else if ((count / 2) >= DTPI_MAX_FACTOR) {
+        i = DTPI_MAX_FACTOR + 1;
+    }
+
+    while ((count % --i) != 0 && i > 1);
+
+    return i;
+}
+
+/*
+ * DTP_pool_create - create a new datatype pool with specified signature
+ *
+ * basic_type:        signature's basic datatype
+ * basic_type_count:  signature's basic datatype count
+ * dtp:               returned datatype pool object
+ */
+int DTP_pool_create(MPI_Datatype basic_type, MPI_Aint basic_type_count, DTP_t * dtp)
+{
+    int i, err = DTP_SUCCESS;
+    int num_objs;
+    int env_num_objs = INT_MAX;
+    char *env_num_objs_str = NULL;
+    struct DTP_obj_array_s *obj_array = NULL;
+
+    DTPI_OBJ_ALLOC_OR_FAIL(*dtp, sizeof(**dtp));
+
+    (*dtp)->DTP_pool_type = DTP_POOL_TYPE__BASIC;
+
+    /* get number of objects from environment */
+    if ((env_num_objs_str = getenv("DTP_NUM_OBJS"))) {
+        env_num_objs = atoi(env_num_objs_str);
+    }
+
+    if (DTPI_Get_max_fact(basic_type_count) == 1) {
+        num_objs = (env_num_objs < DTPI_OBJ_LAYOUT_SIMPLE__NUM && env_num_objs > 0) ?
+            env_num_objs : DTPI_OBJ_LAYOUT_SIMPLE__NUM;
+    } else {
+        num_objs =
```

Comment:
I do not understand the strategy that switches between `DTPI_OBJ_LAYOUT_LARGE__NUM` and `DTPI_OBJ_LAYOUT_SIMPLE__NUM`. Why do you check `DTPI_Get_max_fact(basic_type_count)` ?